src/python/pants/backend/go/target_type_rules.py

```diff
@@ -24,8 +46,103 @@ async def inject_go_module_dependency(request: InjectGoModuleDependency) -> Inje
         return InjectedDependencies()
 
 
+class InferGoPackageDependenciesRequest(InferDependenciesRequest):
+    infer_from = GoPackageSources
+
+
+@rule
+async def infer_go_dependencies(
+    request: InferGoPackageDependenciesRequest,
+    goroot_imports: ResolvedImportPathsForGoLangDistribution,
+) -> InferredDependencies:
+    this_go_package = await Get(
+        ResolvedGoPackage, ResolveGoPackageRequest(request.sources_field.address)
+    )
+
+    # Obtain all go_package targets under this package's go_module.
+    assert this_go_package.module_address is not None
+    spec_path = this_go_package.module_address.spec_path
+    address_specs = [
+        MaybeEmptySiblingAddresses(spec_path),
+        MaybeEmptyDescendantAddresses(spec_path),
+    ]
+    candidate_targets = await Get(UnexpandedTargets, AddressSpecs(address_specs))
+    go_package_targets = [
+        tgt
+        for tgt in candidate_targets
+        if tgt.has_field(GoPackageSources)
+        and not tgt.address.is_file_target
+        and tgt.address != this_go_package.address
+    ]
+
+    # Find all go_external_modules in the repo and map their import paths to their address.
+    candidate_go_external_module_targets = await Get(
+        UnexpandedTargets, AddressSpecs([MaybeEmptyDescendantAddresses("")])
+    )
+    go_external_module_targets = [
+        tgt for tgt in candidate_go_external_module_targets if isinstance(tgt, GoExternalModule)
+    ]
+    third_party_import_path_to_address = {
+        tgt[GoImportPath].value: tgt.address for tgt in go_external_module_targets
+    }
+
+    # Resolve all of the packages found.
+    first_party_import_path_to_address = {}
```

Comment:
Thanks to source roots, is it theoretically possible to have 2 Go targets with the same path? In Python land, this looks like having `src/py/utils.py` and `tests/py/utils.py` where `src/py` and `tests/py` are source roots.

If so, you also need `ambiguous_first_party_import_path_to_addresses: dict[str, set[Address]]`. You'll need this so that you can warn when ambiguous imports are detected.

I think the Python Protobuf Module Mapping code is probably the most direct comparison:

https://github.com/pantsbuild/pants/blob/650a9b3cd26bc48751e2c9e570699b6779dbfc0b/src/python/pants/backend/codegen/protobuf/python/python_protobuf_module_mapper.py#L31-L73

--

It's fine to punt on handling ambiguity, but it is one of the most important details for dep inference to handle correctly and it will change how some of this code is factored. Might be better to do directly, but fine w/ me how you want to prioritize.

Reply:
They should have separate import paths because they would have to be owned by separate `go_module` targets because we require an owning `go_module` and they could not be in the same `go_module` given the different source roots.

Reply:
For now, I'm going to punt and just make the minor edits to not fail dep inference if an import path cannot be resolved. I want to get third party code building. The big open issue with this is how to handle vendored third party code. I need to figure out how that will work before refactoring this code.